source/common/grpc/google_grpc_utils.cc

```diff
@@ -0,0 +1,111 @@
+#include "common/grpc/google_grpc_utils.h"
+
+#include <atomic>
+#include <cstdint>
+#include <cstring>
+#include <string>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+#include "common/common/empty_string.h"
+#include "common/common/enum_to_int.h"
+#include "common/common/fmt.h"
+#include "common/common/macros.h"
+#include "common/common/stack_array.h"
+#include "common/common/utility.h"
+
+#include "absl/strings/match.h"
+
+namespace Envoy {
+namespace Grpc {
+
+struct BufferInstanceContainer {
+  BufferInstanceContainer(int ref_count, Buffer::InstancePtr&& buffer)
+      : ref_count_(ref_count), buffer_(std::move(buffer)) {}
+  std::atomic<uint32_t> ref_count_; // In case gPRC dereferences in a different threads.
+  Buffer::InstancePtr buffer_;
+
+  static void derefBufferInstanceContainer(void* container_ptr) {
+    auto container = reinterpret_cast<BufferInstanceContainer*>(container_ptr);
+    container->ref_count_--;
+    // This is safe because the ref_count_ is never incremented.
+    if (container->ref_count_ <= 0) {
+      delete container;
+    }
+  }
+};
+
+grpc::ByteBuffer GoogleGrpcUtils::makeByteBuffer(Buffer::InstancePtr&& buffer_instance) {
+  if (!buffer_instance) {
+    return {};
+  }
+  Buffer::RawSlice on_raw_slice;
+  // NB: we need to pass in >= 1 in order to get the real "n" (see Buffer::Instance for details).
+  int n_slices = buffer_instance->getRawSlices(&on_raw_slice, 1);
+  if (n_slices <= 0) {
+    return {};
+  }
+  auto container = new BufferInstanceContainer{n_slices, std::move(buffer_instance)};
+  if (n_slices == 1) {
+    grpc::Slice one_slice(on_raw_slice.mem_, on_raw_slice.len_,
+                          &BufferInstanceContainer::derefBufferInstanceContainer, container);
+    return {&one_slice, 1};
+  }
+  STACK_ARRAY(many_raw_slices, Buffer::RawSlice, n_slices);
```

Comment:
We're missing coverage of this, see https://211711-65214191-gh.circle-artifacts.com/0/coverage/coverage.source_common_grpc_google_grpc_utils.cc.html. I would vote not to have the singleton optimization unless it really is shown to be important performance wise, but if you prefer to keep it, then please add coverage for both branches.

Reply:
Score one for coverage.  I had a test for 3 slices but the add() function was combining the data so there was only one.  Updated the test case.